### PR TITLE
daemoncontroller.go: avoid panic

### PR DESCRIPTION
**Pull request by zhangxiaoyu-zidif**
add Comma-ok to avoid panic.

### CODE REVIEW

**Comment by lukaszo** [comment]
@k8s-bot test this

**Comment by lukaszo** [comment]
@zhangxiaoyu-zidif did you hit a panic or is this just a bug prevention?

**Comment by lukaszo** [comment]
@k8s-bot ok to test

**Comment by zhangxiaoyu-zidif** [comment]
@lukaszo I just want to prevent bug now :).  
I am interested in daemonset deployment and plan to apply it to work. I need to review all relative codes to make sure it could works well without some bugs. =)

**Comment by zhangxiaoyu-zidif** [comment]
@lukaszo @janetkuo 
could you approve this PR?

File: pkg/controller/daemon/daemoncontroller.go [hunk 1]
```diff
@@ -317,8 +317,14 @@ func (dsc *DaemonSetsController) addPod(obj interface{}) {
 // up. If the labels of the pod have changed we need to awaken both the old
 // and new set. old and cur must be *v1.Pod types.
 func (dsc *DaemonSetsController) updatePod(old, cur interface{}) {
-	curPod := cur.(*v1.Pod)
-	oldPod := old.(*v1.Pod)
+	curPod, ok := cur.(*v1.Pod)
+	if !ok {
+		return
```
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller.go]: please also log error message if it's failed. ignore is worse than panic, it's hard to admin/dev to investigate issue if any.

Comment (zhangxiaoyu-zidif) [inline pkg/controller/daemon/daemoncontroller.go]: yes, you are right. I will add them soon.

File: pkg/controller/daemon/daemoncontroller.go [hunk 2]
```diff
@@ -317,8 +317,16 @@ func (dsc *DaemonSetsController) addPod(obj interface{}) {
 // up. If the labels of the pod have changed we need to awaken both the old
 // and new set. old and cur must be *v1.Pod types.
 func (dsc *DaemonSetsController) updatePod(old, cur interface{}) {
-	curPod := cur.(*v1.Pod)
-	oldPod := old.(*v1.Pod)
+	curPod, ok := cur.(*v1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("cur is not a pod %#v", cur))
```
Comment (k82cn) [inline pkg/controller/daemon/daemoncontroller.go]: The `HandleError` will also panic :).

**Comment by k82cn** [comment]
> I am interested in daemonset deployment and plan to apply it to work. I need to review all relative codes to make sure it could works well without some bugs. =)

Welcome to k8s :). I think you can also find some "help-wanted" issues which is also a good start for newbee :).

**Comment by zhangxiaoyu-zidif** [comment]
@k82cn Thanks for your reminding and advisement. I will start to deal with issues binding "help-wanted" label.  :)

**Comment by zhangxiaoyu-zidif** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by janetkuo** [comment]
@zhangxiaoyu-zidif thanks a lot for the fix! Such check is generally good but isn't necessary here, because DaemonSet controller uses pod informer which makes sure that the passed in objects are `v1.Pod` already. 

Controller manager calls [`startDaemonSetController`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/cmd/kube-controller-manager/app/extensions.go#L30-L41) and sets informers (including pod informer). The pod informer is created by calling [`NewSharedIndexInformer`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L76-L90) and sets `objectType` to [`v1.Pod`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/staging/src/k8s.io/client-go/informers/core/v1/pod.go#L54). The only exception is `DeleteFunc` which sometimes gets an `cache.DeletedFinalStateUnknown` object.

**Review by janetkuo — CHANGES_REQUESTED** [review]
@zhangxiaoyu-zidif thanks a lot for the fix! Such check is generally good but isn't necessary here, because DaemonSet controller uses pod informer which makes sure that the passed in objects are `v1.Pod` already: 

Controller manager calls [`startDaemonSetController`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/cmd/kube-controller-manager/app/extensions.go#L30-L41) and sets informers (including pod informer). The pod informer is created by calling [`NewSharedIndexInformer`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L76-L90) and sets `objectType` to [`v1.Pod`](https://github.com/kubernetes/kubernetes/blob/v1.7.0-alpha.2/staging/src/k8s.io/client-go/informers/core/v1/pod.go#L54). The only exception is `DeleteFunc` which sometimes gets a `cache.DeletedFinalStateUnknown` object.

**Comment by zhangxiaoyu-zidif** [comment]
@janetkuo Thank you for you explaining. I close it. :)

**Pull request closed without merging**